lib/internal/streams/readable.js

```diff
@@ -1057,6 +1058,22 @@ Readable.prototype.wrap = function(stream) {
   return this;
 };
 
+Readable.prototype.collect = async function(limit = Infinity) {
+  const bufs = [];
+  let len = 0;
+  const isObjectMode = this._readableState.objectMode;
+  const encoding = isObjectMode ? false : this._readableState.encoding;
+  for await (const chunk of this) {
+    const chunkLen = isObjectMode ? 1 : chunk.length;
+    if (len + chunkLen > limit) {
```

Comment:
shouldn't this push back the chunk part that is past the limit for other consumers to use? e.g.

```mjs
let buff = new TextEncoder().encode('Hello world!');
stream.write(buff);
await stream.collect(5); // ~ 'Hello'
await stream.collect(7); // ~ ' world!'
```

I think a test is fine, I'm not 100% familiar with the behavior in async iterator mode. Also, does this need to account for unicode?

Reply:
> shouldn't this push back the chunk part that is past the limit for other consumers to use?

Maybe in a non-rejecting implementation like we're discussing elsehwere in this PR, yeah. Otherwise, "other consumers" don't exist.

What you're doing in the example here seems more like it should happen as an option to [`.iterator()`](https://nodejs.org/dist/latest-v16.x/docs/api/stream.html#stream_readable_iterator_options).

e.g.

```js
let buff = new TextEncoder().encode('Hello world!');
stream.write(buff);
let iter = stream.iterator({ limit: 5 })
(await iter.next()).value; // ~ 'Hello'
(await iter.next()).value; // ~ ' world!'
```

Or an async version of [`.read([size])`](https://nodejs.org/dist/latest-v16.x/docs/api/stream.html#stream_readable_read_size).

> Also, does this need to account for unicode?

I can add a test or two, but I think it doesn't matter? If encoding is set, then `StringDecoder` will do the needful behind the scenes, won't it?

Reply:
@bengl using a stream as a tee/buffered stream is my concern. There isn't a clear guarantee no other listeners are consuming on the same stream or will in the future even if we reject.

Reply:
We can optionally break out of the loop on the limit, allowing other listeners to get future chunks. This would be the `rejectAtLimit: false` option I propose below.

Reply:
so a slight difference is that you need to push back the unused portion of the buffers being read. So if you only partially read "hello world!" for example, the subsequent `.collect()` calls can get the remainder rather than having it be dropped.

Reply:
That should be easy enough with an `.unshift()` before the `break`. I'll do that and add tests as soon as I have a bit of spare time.

Reply:
.. and that was this evening. PTAL at the latest changes.